JIT: Simplify JitDisasm matching behavior and support namespaces/generics in release

jakobbotsch opened the pull request:
This changes how the JIT matches method names and signatures for method
sets (e.g. JitDisasm). It also starts printing method instantiations for full method names
and makes references to types consistent in generic instantiations and the signature.

In addition it starts supporting generic instantiations in release too.

To do this, most of the type printing is moved to the JIT, which also aligns the output
between crossgen2 and the VM (there were subtle differences here, like spaces between generic type arguments).
More importantly, we (for the most part) stop relying on JIT-EE methods that are documented to only be for debug purposes.

The new behavior of the matching is the following:

* The matching behavior is always string based.
* The `JitDisasm` string is a space-separated list of patterns. Patterns can arbitrarily
  contain both '*' (match any characters) and '?' (match any 1 character).
* The string matched against depends on characters in the pattern:
  + If the pattern contains a ':' character, the string matched against is prefixed by the class name and a colon
  + If the pattern contains a '(' character, the string matched against is suffixed by the signature
  + If the class name (part before colon) contains a '[', the class contains its generic instantiation
  + If the method name (part between colon and '(') contains a '[', the method contains its generic instantiation

For example, consider
```csharp
namespace MyNamespace
{
    public class C<T1, T2>
    {
        [MethodImpl(MethodImplOptions.NoInlining)]
        public void M<T3, T4>(T1 arg1, T2 arg2, T3 arg3, T4 arg4)
        {
        }
    }
}

new C<sbyte, string>().M<int, object>(default, default, default, default); // compilation 1
new C<int, int>().M<int, int>(default, default, default, default); // compilation 2
```

The full strings are:
Before:
```scala
MyNamespace.C`2[SByte,__Canon][System.SByte,System.__Canon]:M(byte,System.__Canon,int,System.__Canon)
MyNamespace.C`2[Int32,Int32][System.Int32,System.Int32]:M(int,int,int,int)
```
Notice no method instantiation and the double class instantiation, which seems like an EE bug. Also two different names are used for `sbyte`: `System.SByte` and `byte`.

After the change the strings are:
```scala
MyNamespace.C`2[byte,System.__Canon]:M[int,System.__Canon](byte,System.__Canon,int,System.__Canon)
MyNamespace.C`2[int,int]:M[int,int](int,int,int,int)
```

The following strings will match both compilations:
```scala
M
*C`2:M
*C`2[*]:M[*](*)
MyNamespace.C`2:M
```
The following will match only the first one:
```scala
M[int,*Canon]
MyNamespace.C`2[byte,*]:M
M(*Canon)
```

There is one significant change in behavior here, which is that I have removed the special case that allows matching class names without namespaces. In particular, today `Console:WriteLine` would match all overloads of `System.Console.WriteLine`, while after this change it will not match. However, with generalized wild cards the replacement is simple in `*Console:WriteLine`.

jakobbotsch committed:
Remove SPILLED JitDisasm in favor of JitDisasmSpilled

jakobbotsch committed:
JIT: Simplify MethodSet matching behavior

This changes how the JIT matches method names and signatures for method
sets (e.g. JitDisasm). The new behavior is the following:

* The matching behavior is always string based.
* The JitDisasm string is a space-separated list of patterns. Patterns
  can be quoted if they need to include spaces, and can arbitrarily
  contain both * (match any characters) and ? (match any 1 character).
* If the pattern contains a : character, the string matched against is
  prefixed by "ClassName:".
* If the pattern contains a ( character then the string matched against
  ends with its signature in the usual format.
* Type names do not include namespaces.
* The signature matched against does not contain return types or
  "is-instance" specifiers.

For example, *:WriteLine is equivalent to WriteLine and will match all
methods named WriteLine. Console:WriteLine will match all overloads of
WriteLine defined in a class called Console. Console:WriteLine(String)
and WriteLine(String) will match only the string overload.

jakobbotsch commented:
@dotnet/jit-contrib, any concerns about this change in behavior of the matching?
Also cc @stephentoub

jakobbotsch commented on src/coreclr/jit/jitconfig.cpp:
```diff
@@ -320,100 +92,90 @@ void JitConfigValues::MethodSet::destroy(ICorJitHost* host)
-    int numArgs = sigInfo != nullptr ? sigInfo->numArgs : -1;
+    // names[hasClassName][hasSignature]
+    char*     names[2][2] = {};
+    Compiler* comp        = JitTls::GetCompiler();
```
Are we worried about this? It should be a single call in release builds and only when `JitDisasm` is non-empty.
We need it to allocate memory in rare cases where the string we are matching against is longer than 1024 characters.

jakobbotsch committed:
Update docs, minor nits

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into jit-method-name-string-match

AaronRobinsonMSFT commented on docs/design/coreclr/jit/viewing-jit-dumps.md:
```diff
@@ -120,29 +120,31 @@ These can be set in one of three ways:
+Some environment variables such as `COMPlus_JitDump` take a set of patterns specifying method names. The matching works in the following way:
+* The environment variable is a space-separated list of patterns that can be quoted if they need to contain spaces.
+* Patterns can contain * and ? wildcards matching respectively any characters and any 1 character.
+* If the pattern contains a : character, then the string matched against is prefixed with `ClassName:`.
```
```suggestion
* If the pattern contains a ':' character, then the string matched against is prefixed with `ClassName:`.
```

AaronRobinsonMSFT commented on docs/design/coreclr/jit/viewing-jit-dumps.md:
```diff
@@ -120,29 +120,31 @@ These can be set in one of three ways:
+* The environment variable is a space-separated list of patterns that can be quoted if they need to contain spaces.
+* Patterns can contain * and ? wildcards matching respectively any characters and any 1 character.
+* If the pattern contains a : character, then the string matched against is prefixed with `ClassName:`.
+* If the patter ncontains a ( character, then the string matched against is suffixed with its signature.
```
```suggestion
* If the pattern contains a '(' character, then the string matched against is suffixed with its signature.
```

AaronRobinsonMSFT commented on docs/design/coreclr/jit/viewing-jit-dumps.md:
```diff
@@ -120,29 +120,31 @@ These can be set in one of three ways:
 ## Specifying method names
 
-The complete syntax for specifying a single method name (for a flag that takes a method name, such as `COMPlus_JitDump`) is:
+Some environment variables such as `COMPlus_JitDump` take a set of patterns specifying method names. The matching works in the following way:
```
```suggestion
Some environment variables such as `DOTNET_JitDump` take a set of patterns specifying method names. The matching works in the following way:
```

jakobbotsch committed:
Apply suggestions from code review

Co-authored-by: Aaron Robinson <arobins@microsoft.com>

BruceForstall commented:
> any concerns about this change in behavior of the matching?

Can you give examples of the change in behavior? It's not clear to me from the description, with the exception of supporting matching arguments, which is a great improvement.

> * If the pattern contains a : character, the string matched against is prefixed by "ClassName:".

What about the ":return_type" and ":this" suffixes we currently use?

BruceForstall commented on docs/design/coreclr/jit/viewing-jit-dumps.md:
```diff
@@ -120,29 +120,31 @@ These can be set in one of three ways:
 ## Specifying method names
 
-The complete syntax for specifying a single method name (for a flag that takes a method name, such as `COMPlus_JitDump`) is:
+Some environment variables such as `DOTNET_JitDump` take a set of patterns specifying method names. The matching works in the following way:
```
At some point (but not in this change) we should probably make a complete change of our documentation to use the DOTNET_ form instead of the COMPlus_ form, but it's odd to only change this one case in this file.

jakobbotsch replied:
Sure, I'll revert this.

```suggestion
Some environment variables such as `COMPlus_JitDump` take a set of patterns specifying method names. The matching works in the following way:
```

jakobbotsch commented:
> Can you give examples of the change in behavior? It's not clear to me from the description, with the exception of supporting matching arguments, which is a great improvement.

To be honest it is not clear to me what the full extent of the existing parser's behavior is. One difference is that it allows `ClassName::MethodName` (two colons) -- we now only allow one colon. As you say, another one is in the matching of parameters. A final one is the treatment of wildcards, which I never really understood.

At least the way I normally have used method sets the behavior should not have changed, but I'm not sure if other people on the team are using some more advanced features.

> What about the ":return_type" and ":this" suffixes we currently use?

It is not included as part of the signature that the matching works against.

MichalPetryka commented:
Do those changes allow the filter to be used with generic types?

jakobbotsch commented:
> Do those changes allow the filter to be used with generic types and methods?

Not in release builds. It probably needs some work on the JIT-EE interface so it won't happen as part of this change.

jakobbotsch committed:
Update docs/design/coreclr/jit/viewing-jit-dumps.md

jakobbotsch committed:
Use appendClassName to get namespaces and generic instantiations

jakobbotsch committed:
Format instantiations in JIT

Like signatures, format the instantiations in the JIT so we get
consistent type names in the entire string.

jakobbotsch committed:
Minor adjustments

jakobbotsch committed:
Add method instantiations

jakobbotsch committed:
Update docs

jakobbotsch committed:
Go back to square brackets

jakobbotsch committed:
Congruence

jakobbotsch committed:
Single quotes

jakobbotsch committed:
Add some SPMI error traps

jakobbotsch committed:
Switch vsnprintf_s -> _vsnprintf_s, move StringPrinter::Printf to cpp file

jakobbotsch committed:
Another compilation fix

jakobbotsch committed:
Change crossgen2/ILC to fill in instantiations always

While there as a comment that this is expensive, in my measuring of
crossgenning SPC I could not see any performance difference.

jakobbotsch commented on src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -776,27 +776,25 @@ private void Get_CORINFO_SIG_INFO(MethodDesc method, CORINFO_SIG_INFO* sig, Meth
+            if (method.Instantiation.Length != 0)
+            {
+                sig->sigInst.methInst = GetJitInstantiation(method.Instantiation);
+            }
```
I have moved this out of the `method.IsIntrinsic` path to unconditionally fill it, since it is used by JIT for `JitDisasm` now and making assumptions on the exact scenarios this is used by the JIT is a bit icky.

If we are worried about performance (I did not see any measurable difference in crossgen2 times of SPC) then I think we should remove `CORINFO_SIG_INFO::sigInst` and put this on the same plan as `ICorStaticInfo::getTypeInstantiationArgument`. But I think we can wait and see the crossgen2 TP graphs in PowerBI.

jakobbotsch committed:
Fix a couple more angle brackets, add some braces to single line ifs

jakobbotsch committed:
Make globbing quadratic instead of exponential

jakobbotsch committed:
Use asCorInfoType instead of getTypeForPrimitiveValueClass

jakobbotsch committed:
Run jit-format

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into jit-method-name-string-match

jakobbotsch committed:
SPMI: Fix getTypeInstantiationArgument

This was completely ignoring the index.

jakobbotsch commented on src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp:
```diff
@@ -6461,27 +6461,33 @@ const char* MethodContext::repGetClassNameFromMetadata(CORINFO_CLASS_HANDLE cls,
+                                                    CORINFO_CLASS_HANDLE result)
 {
     if (GetTypeInstantiationArgument == nullptr)
-        GetTypeInstantiationArgument = new LightWeightMap<DWORDLONG, DWORDLONG>();
+        GetTypeInstantiationArgument = new LightWeightMap<DLD, DWORDLONG>();
 
-    DWORDLONG key = CastHandle(cls);
+    DLD key;
+    ZeroMemory(&key, sizeof(key));
+    key.A = CastHandle(cls);
+    key.B = index;
     DWORDLONG value = CastHandle(result);
     GetTypeInstantiationArgument->Add(key, value);
     DEBUG_REC(dmpGetTypeInstantiationArgument(key, value));
 }
-void MethodContext::dmpGetTypeInstantiationArgument(DWORDLONG key, DWORDLONG value)
+void MethodContext::dmpGetTypeInstantiationArgument(DLD key, DWORDLONG value)
 {
-    printf("GetTypeInstantiationArgument key - classNonNull-%llu, value NonNull-%llu", key, value);
+    printf("GetTypeInstantiationArgument key - classNonNull-%llu, index-%u, value NonNull-%llu", key.A, key.B, value);
     GetTypeInstantiationArgument->Unlock();
 }
 CORINFO_CLASS_HANDLE MethodContext::repGetTypeInstantiationArgument(CORINFO_CLASS_HANDLE cls, unsigned index)
 {
     CORINFO_CLASS_HANDLE result = nullptr;
 
-    DWORDLONG key = CastHandle(cls);
+    DLD key;
+    ZeroMemory(&key, sizeof(key));
+    key.A = CastHandle(cls);
+    key.B = index;
```
@BruceForstall annoyingly, I suppose this bug fix means I should bump the JIT-EE GUID in this change even though there are no changes to the JIT-EE interface.

BruceForstall replied:
True; any time the SPMI binary format changes, we also need a JIT-EE GUID change.

jakobbotsch committed:
Bump JIT-EE GUID

jakobbotsch committed:
Add header comments, rename some parameters

jakobbotsch commented:
@dotnet/jit-contrib This should be ready now. I have updated the original text of the PR, so please read it and give me your thoughts on that. This behavior change probably being the largest thing:
> There is one significant change in behavior here, which is that I have removed the special case that allows matching class names without namespaces. In particular, today `Console:WriteLine` would match all overloads of `System.Console.WriteLine`, while after this change it will not match. However, with generalized wild cards the replacement is simple in `*Console:WriteLine`.

It would be nice to get a review on this today so I can hopefully merge this on the weekend and avoid disruption due to the JIT-EE GUID change.

jakobbotsch commented on src/coreclr/jit/jitconfigvalues.h:
```diff
@@ -190,10 +190,11 @@ CONFIG_STRING(JitDisasmAssemblies, W("JitDisasmAssemblies")) // Only show JitDis
-CONFIG_INTEGER(JitDumpTier0, W("JitDumpTier0"), 1)                     // Dump tier0 requests
-CONFIG_INTEGER(JitDumpAtOSROffset, W("JitDumpAtOSROffset"), -1)        // Only dump OSR requests for this offset
-CONFIG_INTEGER(JitDumpInlinePhases, W("JitDumpInlinePhases"), 1)       // Dump inline compiler phases
+CONFIG_INTEGER(JitDisasmSpilled, W("JitDisasmSpilled"), 0)      // Display native code when any register spilling occurs
```
This was shoehorned into `JitDisasm` by specifying "SPILLED" before. I have introduced a separate variable for it instead.

BruceForstall commented on src/coreclr/inc/corinfo.h:
```diff
@@ -2893,9 +2893,15 @@ class ICorStaticInfo
-    // and if 'moduleName' is non-null, it sets it to something that will
-    // says which method (a class name, or a module name)
+    // This function returns the method name and if 'moduleName' is non-null,
+    // it sets it to something that will says what contains the method (a class
+    // name, or a module name). Note that the moduleName parameter is for
```
grammar
```suggestion
    // it sets it to something that contains the method (a class
    // name, or a module name). Note that the moduleName parameter is for
```

Also, what does it mean "for diagnostics only"?

jakobbotsch replied:
In the EE, `moduleName` is set to basically "what it can get its hands on": in debug builds it has a debug-only buffer available that it uses to create a namespace qualified class name. But in release builds it does not have any buffer available, so it just returns a pointer to the class name only. So using `moduleName` from the JIT introduces different behavior depending on whether coreclr.dll is debug or release.

This PR stops using this parameter and instead uses `getClassNameFromMetadata` to create the namespace qualified names inside the JIT. We still use `getMethodName`, but only the returned method name, which seems to not have this "debug-only/diagnostic-only" flavor to it, so I have updated the documentation to try to make this more clear.

BruceForstall commented on docs/design/coreclr/jit/viewing-jit-dumps.md:
```diff
@@ -120,29 +120,59 @@ These can be set in one of three ways:
-The complete syntax for specifying a single method name (for a flag that takes a method name, such as `COMPlus_JitDump`) is:
-
+Some environment variables such as `COMPlus_JitDump` take a set of patterns specifying method names. The matching works in the following way:
+* The JitDisasm string is a space-separated list of patterns. Patterns can arbitrarily contain both '*' (match any characters) and '?' (match any 1 character).
```
nit
```suggestion
* A method set string is a space-separated list of patterns. Patterns can arbitrarily contain both '*' (match any characters) and '?' (match any 1 character).
```

BruceForstall reviewed (approved):
This looks really great!

I only pointed out a couple nits, but feel free to ignore.

jakobbotsch committed:
Apply suggestions from code review

Co-authored-by: Bruce Forstall <brucefo@microsoft.com>

jakobbotsch commented:
superpmi failures are expected due to the JIT-EE GUID change.

jakobbotsch commented:
/backport to release/7.0
